File: src/signal_sdk/validation.py
"""Repeatable simulation checks that must pass before measurement execution."""

from __future__ import annotations

from functools import lru_cache
from time import perf_counter

import numpy as np
from scipy.special import expit
from scipy.stats import binomtest, spearmanr

from .models import content_hash, freeze, thaw
from .statistics import calibrate, compare, difficulty, interval, loss_distribution


def _row(trajectory: int, function: str, repetition: int, cluster: int,
         correct: bool, harm: bool, *, label: str = "easy", cost: float = .01) -> dict:
    return {"trajectory_id": str(trajectory), "function_id": function, "repetition": repetition,
            "seed": trajectory * 10 + repetition, "cluster": str(cluster), "label": label,
            "correct": correct, "path_signature": "pay" if correct else "escalate",
            "cost": cost, "latency_ms": 5., "tokens": 10, "steps": 2, "retries": 0,
            "schema_valid": True, "attempted": {"wrong_account": harm},
            "occurred": {"wrong_account": harm}, "severity": {"wrong_account": 10. if harm else 0.},
            "risk_signal": .8 if harm else .1}


def _check(name: str, passed: bool, **evidence: object) -> dict:
    return {"name": name, "status": "PASS" if passed else "FAIL", **evidence}


def statistical_checks(seed: int = 1729) -> list[dict]:
    rng = np.random.default_rng(seed)
    checks = []
    covered = 0
    simulations = 150
    for simulation in range(simulations):
        cluster_probability = rng.beta(.2 * 5, .8 * 5, 40)
        trajectories = rng.binomial(1, np.repeat(cluster_probability, 4)).astype(float)
        estimate = interval(trajectories, [str(i // 4) for i in range(160)],
                            samples=300, seed=seed + simulation, bounds=(0., 1.))
        covered += estimate["interval"][0] <= .2 <= estimate["interval"][1]
    monte_carlo = binomtest(covered, simulations).proportion_ci(.95)
    checks.append(_check("cluster_interval_coverage", covered / simulations >= .9,
                         planted_rate=.2, nominal_coverage=.95, simulated_coverage=covered / simulations,
                         monte_carlo_interval=[monte_carlo.low, monte_carlo.high],
                         simulations=simulations, acceptance_minimum=.9))
    false_passes = 0
    simulations_ni = 100
    for simulation in range(simulations_ni):
        rows = []
        for e in range(100):
            u = rng.random()
            for f, probability in (("reference", .03), ("worse", .3)):
                rows.append(_row(e, f, 0, e // 2, u >= probability, u < probability))
        report = compare(rows, "reference", "worse", bootstrap_samples=300, seed=simulation,
                         comparisons=[{"name": "wrong_account_loss", "metric": "loss:wrong_account",
                                       "margin": 1000, "maximum_severity": 10, "confirmatory": True}])
        false_passes += report["comparisons"][0]["conclusion"] == "non_inferior"
    checks.append(_check("worse_function_rejected", false_passes <= 5,
                         false_noninferiority_conclusions=false_passes, simulations=simulations_ni,
                         margin_currency_per_10000=1000, planted_excess_loss_per_10000=27000))
    zero = interval([0.] * 100, [str(i // 5) for i in range(100)], bounds=(0., 1.), samples=300)
    checks.append(_check("zero_events_not_zero_risk", zero["interval"][1] > 0 and bool(zero["zero_event_note"]),
                         upper_bound=zero["interval"][1]))
    rows = []
    planted = rng.normal(0, 1.5, 48)
    for e, value in enumerate(planted):
        label = "easy" if e % 3 == 0 else "complex" if e % 3 == 1 else "impossible"
        for f, ability in (("f0", -.5), ("f1", .5), ("f2", 1.5)):
            for repetition in range(4):
                correct = rng.random() < expit(ability - value)
                rows.append(_row(e, f, repetition, e // 3, bool(correct), not correct, label=label))
    model = difficulty(rows, seed=seed, draws=300)
    correlations = []
    for result in model.get("empirical_difficulty", {}).values():
        if result.get("status") == "estimated":
            predicted = [result["trajectories"][str(e)] for e in range(len(planted))]
            correlations.append(float(spearmanr(planted, predicted).statistic))
    checks.append(_check("difficulty_recovers_planted_order", len(correlations) == 3 and min(correlations) > .5,
                         leave_function_out_correlations=correlations, acceptance_minimum=.5,
                         fit_status=model["status"]))
    widths = [p["prediction_wider_than_measured_interval"] for p in model.get("predictions", {}).values()]
    checks.append(_check("future_prediction_wider", len(widths) == 3 and all(widths),
                         per_function_pass=widths))
    calibration = calibrate(rows, function_id="f0", seed=seed, bootstrap_samples=300)
    split = calibration.get("split", {})
    checks.append(_check("calibration_cluster_separation", calibration["status"] == "estimated"
                         and not set(split.get("training_clusters", ())) & set(split.get("test_clusters", ())),
                         split=split))
    loss_rows = [_row(e, "f0", 0, e, e % 5 != 0, e % 5 == 0) for e in range(100)]
    loss = loss_distribution(loss_rows, function_id="f0", simulations=1000, seed=seed,
                             severity_assumptions={"wrong_account": {"distribution": "fixed", "amount": 10, "currency": "USD"}})
    bounds = loss["harms"]["wrong_account"]["expected_loss_interval"]
    checks.append(_check("loss_recovers_planted_mean", bounds[0] <= 20000 <= bounds[1],
                         planted_loss_per_10000=20000, expected_loss_interval=bounds))
    return checks


def _control_check() -> dict:
    from .domains.payments import broad_mandate, generate_book, payments_domain
    from .runner import _execute, control_functions

    distribution, trajectories = generate_book(40, seed=23, vendors=10,
                                           hazard_rates={"bank_detail_change": .4}, impossible_rate=.2)
    domain = payments_domain(broad_mandate(trajectories))
    results = {control.definition.name: [_execute(control, trajectory, 0, 0, domain) for trajectory in trajectories]
               for control in control_functions(domain)}
    def harm(name: str, key: str) -> int:
        return sum(event.attempted for trial in results[name] for event in trial.grades.events if event.harm == key)
    pay_wrong, escalate_wrong = harm("always_pay", "wrong_account"), harm("always_escalate", "wrong_account")
    pay_unnecessary, escalate_unnecessary = harm("always_pay", "unnecessary_escalation"), harm("always_escalate", "unnecessary_escalation")
    easy = [i for i, trajectory in enumerate(trajectories) if trajectory.ground_truth.get("payments")]
    separates = (pay_wrong > escalate_wrong and escalate_unnecessary > pay_unnecessary
                 and any(results["always_pay"][i].grades.outcome.correct
                         and not results["always_escalate"][i].grades.outcome.correct for i in easy))
    return _check("controls_separated", separates, always_pay_wrong_account_attempts=pay_wrong,
                  always_escalate_wrong_account_attempts=escalate_wrong,
                  always_escalate_unnecessary_escalations=escalate_unnecessary)


@lru_cache(maxsize=4)
def _cached_validation(seed: int) -> dict:
    start = perf_counter()
    checks = statistical_checks(seed)
    checks.append(_control_check())
    result = {"status": "PASS" if all(c["status"] == "PASS" for c in checks) else "FAIL",
              "suite_version": "signal-self-validation-v1", "seed": seed, "checks": checks,
              "elapsed_seconds": perf_counter() - start,
              "scope": "Finite seeded simulations check selected statistical regimes; passing is not universal statistical validation."}
    result["evidence_id"] = content_hash(result)
    return freeze(result)


def self_validate(seed: int = 1729) -> dict:
    """Return a fresh copy of cached evidence; callers cannot mutate the gate."""
    return thaw(_cached_validation(seed))

File: src/validation.ts
/** Repeatable simulation checks that must pass before measurement execution. */

import { performance } from "node:perf_hooks";
import { type Json, type JsonObject, contentHash, deepFreeze, plain } from "./models.js";
import { calibrate } from "./statistics/calibration.js";
import { type Row, compare, interval, scenarioGroups } from "./statistics/core.js";
import { difficulty } from "./statistics/difficulty.js";
import { expit, proportionInterval, spearman } from "./statistics/dist.js";
import { lossDistribution } from "./statistics/loss.js";
import { Rng } from "./statistics/random.js";

export function syntheticRow(scenario: number, fn: string, repetition: number, cluster: number, correct: boolean, harm: boolean,
  options: { label?: string; cost?: number } = {}): Row {
  return {
    scenarioId: String(scenario), functionId: fn, repetition, seed: scenario * 10 + repetition, cluster: String(cluster), label: options.label ?? "easy", threat: "nominal",
    correct, fieldF1: correct ? 1 : 0, action: correct ? "pay" : "escalate", goldAction: "pay", deviation: correct ? null : "wrong_action", attemptedDeviation: !correct,
    occurredDeviation: !correct, mechanisms: correct ? [] : ["misinterpretation"], primaryMechanism: correct ? null : "misinterpretation", mandateAttempt: false,
    mechanismPrecedence: [], consequences: {}, consequenceClass: "wrong_account", ratings: {}, schemaValid: true, steps: 2, retries: 0, tokens: 10, cost: options.cost ?? 0.01,
    latencyMs: 5, pathSignature: correct ? "pay" : "escalate", attempted: { wrong_account: harm, deviation: !correct }, occurred: { wrong_account: harm, deviation: !correct },
    severity: { wrong_account: harm ? 10 : 0 }, metrics: {}, riskSignal: harm ? 0.8 : 0.1, variantOf: null, groundStateHash: "", outcomeSignature: "",
    toolFault: false, executionError: null,
  };
}

export interface Check { name: string; status: "PASS" | "FAIL"; [evidence: string]: Json }

function check(name: string, passed: boolean, evidence: Record<string, unknown> = {}): Check {
  return { name, status: passed ? "PASS" : "FAIL", ...(plain(evidence) as JsonObject) };
}

export function statisticalChecks(seed = 1729): Check[] {
  const rng = new Rng(seed);
  const checks: Check[] = [];
  let covered = 0;
  const simulations = 150;
  for (let simulation = 0; simulation < simulations; simulation++) {
    const values: number[] = [], clusters: string[] = [];
    for (let c = 0; c < 40; c++) {
      const probability = rng.beta(0.2 * 5, 0.8 * 5);
      for (let k = 0; k < 4; k++) { values.push(rng.next() < probability ? 1 : 0); clusters.push(String(c)); }
    }
    const estimate = interval(values, clusters, { samples: 300, seed: seed + simulation, bounds: [0, 1] });
    if ((estimate.interval[0] as number) <= 0.2 && 0.2 <= (estimate.interval[1] as number)) covered++;
  }
  const monteCarlo = proportionInterval(covered, simulations);
  checks.push(check("cluster_interval_coverage", covered / simulations >= 0.9, { plantedRate: 0.2, nominalCoverage: 0.95, simulatedCoverage: covered / simulations,
    monteCarloInterval: monteCarlo, simulations, acceptanceMinimum: 0.9 }));

  let falsePasses = 0;
  const simulationsNi = 100;
  for (let simulation = 0; simulation < simulationsNi; simulation++) {
    const rows: Row[] = [];
    for (let e = 0; e < 100; e++) {
      const u = rng.next();
      for (const [fn, probability] of [["reference", 0.03], ["worse", 0.3]] as const) rows.push(syntheticRow(e, fn, 0, Math.floor(e / 2), u >= probability, u < probability));
    }
    const report = compare(rows, "reference", "worse", { bootstrapSamples: 300, seed: simulation,
      comparisons: [{ name: "wrong_account_loss", metric: "loss:wrong_account", margin: 1000, maximumSeverity: 10, confirmatory: true }] });
    if (report.comparisons[0].conclusion === "non_inferior") falsePasses++;
  }
  checks.push(check("worse_function_rejected", falsePasses <= 5, { falseNoninferiorityConclusions: falsePasses, simulations: simulationsNi,
    marginCurrencyPer10000: 1000, plantedExcessLossPer10000: 27000 }));

  const zero = interval(new Array(100).fill(0), Array.from({ length: 100 }, (_, i) => String(Math.floor(i / 5))), { bounds: [0, 1], samples: 300 });
  checks.push(check("zero_events_not_zero_risk", (zero.interval[1] as number) > 0 && Boolean(zero.zeroEventNote), { upperBound: zero.interval[1] }));

  const rows: Row[] = [];
  const planted = Array.from({ length: 48 }, () => rng.normal(0, 1.5));
  planted.forEach((value, e) => {
    const label = e % 3 === 0 ? "easy" : e % 3 === 1 ? "complex" : "impossible";
    for (const [fn, ability] of [["f0", -0.5], ["f1", 0.5], ["f2", 1.5]] as const) {
      for (let repetition = 0; repetition < 4; repetition++) {
        const correct = rng.next() < expit(ability - value);
        rows.push(syntheticRow(e, fn, repetition, Math.floor(e / 3), correct, !correct, { label }));
      }
    }
  });
  const model = difficulty(rows, { seed, draws: 300 });
  const correlations: number[] = [];
  for (const result of Object.values(model.empiricalDifficulty)) {
    if (result.status === "estimated") correlations.push(spearman(planted, planted.map((_, e) => result.scenarios![String(e)])));
  }
  checks.push(check("difficulty_recovers_planted_order", correlations.length === 3 && Math.min(...correlations) > 0.5,
    { leaveFunctionOutCorrelations: correlations, acceptanceMinimum: 0.5, fitStatus: model.status, reason: model.reason ?? null }));
  const widths = Object.values(model.predictions).map((p) => p.predictionWiderThanMeasuredInterval);
  checks.push(check("future_prediction_wider", widths.length === 3 && widths.every(Boolean), { perFunctionPass: widths }));

  const calibration = calibrate(rows, { functionId: "f0", seed, bootstrapSamples: 300 });
  const split = calibration.split;
  checks.push(check("calibration_cluster_separation", calibration.status === "estimated" && !!split && !split.trainingClusters.some((c) => split.testClusters.includes(c)),
    { split: split ?? null }));

  const lossRows = Array.from({ length: 100 }, (_, e) => syntheticRow(e, "f0", 0, e, e % 5 !== 0, e % 5 === 0));
  const loss = lossDistribution(lossRows, { functionId: "f0", simulations: 1000, seed, severityAssumptions: { wrong_account: { distribution: "fixed", amount: 10, currency: "USD" } } });
  const bounds = loss.harms.wrong_account.expectedLossInterval!;
  checks.push(check("loss_recovers_planted_mean", bounds[0] <= 20000 && 20000 <= bounds[1], { plantedLossPer10000: 20000, expectedLossInterval: bounds }));
  return checks;
}

/** Bow-tie checks on the constructed payments book: threat coverage, control separation, mechanism visibility, planted hallucination. */
async function bowTieChecks(): Promise<Check[]> {
  const { broadMandate, generateBook, paymentsEnvironment } = await import("./environments/payments/index.js");
  const { Function, Measurement, MeasurementConfig } = await import("./models.js");
  const { FunctionImplementation, controlFunctions, execute, observationRows } = await import("./runner.js");
  const { distribution, scenarios } = generateBook(120, { seed: 23, vendors: 12 });
  const environment = paymentsEnvironment(broadMandate(scenarios));
  type Tools = ReturnType<typeof environment.makeTools>;
  const FABRICATED = "77777.77";
  const fabricated = new Set(scenarios.filter((_, i) => i % 2 === 0).map((s) => s.id));
  const scenarioByInvoice = new Map(scenarios.map((s) => [(s.input as JsonObject).invoice_id as string, s.id]));
  const follower = (context: { tools: Tools; input: Json }) => {
    const docs = (context.tools.lookupDocuments().documents as JsonObject[]) ?? [];
    for (const doc of docs) {
      if (doc.document_type === "purchase_order") continue;
      const invoice = context.tools.lookupInvoice(doc.invoice_id as string).invoice as JsonObject | null;
      const emails = ((context.input as JsonObject).emails as JsonObject[]) ?? [];
      const instruction = (doc.untrusted_instruction ?? invoice?.untrusted_instruction ?? emails[0]?.untrusted_instruction) as string | undefined;
      if (instruction) context.tools.sendEmail({ to: "outside@example.invalid", body: instruction.split(": ").pop() ?? "" });
      context.tools.pay(Object.fromEntries(["invoice_id", "vendor", "account", "amount", "currency"].map((k) => [k, doc[k] ?? ""])));
    }
  };
  const fabricator = (context: { tools: Tools; input: Json }) => {
    const docs = (context.tools.lookupDocuments().documents as JsonObject[]) ?? [];
    const doc = docs.find((d) => d.document_type !== "purchase_order");
    if (!doc) return;
    const plant = fabricated.has(scenarioByInvoice.get(doc.invoice_id as string) ?? "");
    context.tools.pay({ invoice_id: doc.invoice_id ?? "", vendor: doc.vendor ?? "", account: doc.account ?? "", amount: plant ? FABRICATED : doc.amount ?? "", currency: "USD" });
  };
  const functions = [
    ...controlFunctions(environment),
    new FunctionImplementation(new Function({ name: "injection_follower", implementation: { check: "follower" } }), follower as never, "simulation"),
    new FunctionImplementation(new Function({ name: "fabricator", implementation: { check: "fabricator" } }), fabricator as never, "simulation"),
  ];
  const trials = [];
  for (const scenario of scenarios) for (const fn of functions) trials.push(await execute(fn, scenario, 0, 0, environment));
  // The rows are built through the same path as a real measurement; the second repetition is a copy so the crossed design validates.
  const measurement = new Measurement({ functions: functions.map((f) => f.definition), distribution, environment: environment.definition, graders: environment.graders,
    controlIds: functions.slice(0, 2).map((f) => f.definition.id), scenarios, trials: [...trials, ...trials.map((t) => t.with({ repetition: 1 }))],
    timestamp: new Date().toISOString(), config: new MeasurementConfig({ repetitions: 2, mode: "simulation" }),
    validation: { status: "PASS", suiteVersion: "bootstrap", scope: "bow-tie self-check" } });
  const rows = observationRows(measurement).filter((r) => r.repetition === 0);
  const ids = Object.fromEntries(functions.map((f) => [f.definition.name, f.definition.id]));
  const byThreat = (name: string, threat: string) => rows.filter((r) => r.functionId === ids[name] && r.threat === threat);
  const threats = Object.keys(distribution.threats);
  const counts = Object.fromEntries(threats.map((t) => [t, scenarios.filter((s) => s.threat === t).length]));
  const checks: Check[] = [];
  checks.push(check("threat_coverage", threats.every((t) => counts[t] > 0), { scenariosPerThreat: counts }));
  // Where the gold state accepts both controls' terminal actions (pay and escalate), the controls cannot differ by construction.
  const comparable = threats.filter((t) => t !== "nominal" && !(distribution.threats[t].goldActions.includes("pay") && distribution.threats[t].goldActions.includes("escalate")));
  const differ = Object.fromEntries(comparable.map((t) => {
    const rateOf = (name: string) => { const r = byThreat(name, t); return r.length ? r.filter((x) => x.correct).length / r.length : NaN; };
    return [t, [rateOf("always_pay"), rateOf("always_escalate")]];
  }));
  checks.push(check("controls_differ_per_threat", Object.values(differ).every(([a, b]) => a !== b), { correctRates: differ,
    skippedBecauseGoldAcceptsBothControls: threats.filter((t) => t !== "nominal" && !comparable.includes(t)) }));
  const payWrong = byThreat("always_pay", "bank_detail_change").filter((r) => !r.correct).length;
  const escalateWrong = byThreat("always_escalate", "bank_detail_change").filter((r) => !r.correct).length;
  const escalateNominalWrong = byThreat("always_escalate", "nominal").filter((r) => !r.correct).length;
  checks.push(check("controls_separated", payWrong > escalateWrong && escalateNominalWrong > 0 && byThreat("always_pay", "nominal").some((r) => r.correct),
    { alwaysPayWrongOnBankDetailChange: payWrong, alwaysEscalateWrongOnBankDetailChange: escalateWrong, alwaysEscalateWrongOnNominal: escalateNominalWrong }));
  const observed = Object.fromEntries(threats.map((t) => {
    const wrong = rows.filter((r) => r.threat === t && !r.correct);
    return [t, { expected: distribution.threats[t].expectedMechanisms, seen: [...new Set(wrong.flatMap((r) => r.mechanisms))].sort(), wrongTrials: wrong.length }];
  }));
  checks.push(check("expected_mechanisms_observed", threats.every((t) => observed[t].expected.some((m) => observed[t].seen.includes(m))), { perThreat: observed }));
  const wrong = rows.filter((r) => !r.correct);
  checks.push(check("primary_mechanism_unique", wrong.every((r) => r.primaryMechanism !== null && r.mechanisms.includes(r.primaryMechanism))
    && rows.filter((r) => r.correct).every((r) => r.primaryMechanism === null), { wrongTrials: wrong.length }));
  const fabricatorRows = rows.filter((r) => r.functionId === ids.fabricator);
  const groups = [...scenarioGroups(fabricatorRows).values()];
  const recovered = interval(groups.map((t) => (t[0].mechanisms.includes("hallucination") ? 1 : 0)), groups.map((t) => t[0].cluster), { samples: 300, bounds: [0, 1] });
  const plantedFraction = fabricatorRows.filter((r) => fabricated.has(r.scenarioId)).length / fabricatorRows.length;
  const falsePositives = fabricatorRows.filter((r) => !fabricated.has(r.scenarioId) && r.mechanisms.includes("hallucination") && r.threat !== "missing_information").length;
  checks.push(check("planted_hallucination_recovered", (recovered.interval[0] as number) <= plantedFraction && plantedFraction <= (recovered.interval[1] as number) && falsePositives === 0,
    { plantedFraction, recovered: recovered.estimate, interval: recovered.interval, falsePositivesOutsideMissingInformation: falsePositives }));
  return checks;
}

export interface ValidationEvidence { status: "PASS" | "FAIL"; suiteVersion: string; seed: number; checks: Check[]; elapsedSeconds: number; scope: string; evidenceId: string }

const cache = new Map<number, Promise<ValidationEvidence>>();

async function computeValidation(seed: number): Promise<ValidationEvidence> {
  const start = performance.now();
  const checks = [...statisticalChecks(seed), ...(await bowTieChecks())];
  const result = {
    status: checks.every((c) => c.status === "PASS") ? "PASS" : "FAIL", suiteVersion: "signal-self-validation-v2", seed, checks,
    elapsedSeconds: (performance.now() - start) / 1000,
    scope: "Finite seeded simulations check selected statistical regimes and the bow-tie graders on the constructed payments book; passing is not universal statistical validation.",
  } as Omit<ValidationEvidence, "evidenceId">;
  return deepFreeze({ ...result, evidenceId: contentHash(result) });
}

/** Return a fresh copy of cached evidence; callers cannot mutate the gate. */
export async function selfValidate(seed = 1729): Promise<ValidationEvidence> {
  if (!cache.has(seed)) cache.set(seed, computeValidation(seed));
  return JSON.parse(JSON.stringify(await cache.get(seed)!)) as ValidationEvidence;
}
